// dt-sql-parser/dist/utils/token.js
export var TokenType;
(function (TokenType) {
    /**
     * Enclosed in single/double/back quotation, `` Symbol
     * 'abc', "abc", `abc`
     */
    TokenType["SingleQuotation"] = "SingleQuotation";
    TokenType["DoubleQuotation"] = "DoubleQuotation";
    TokenType["BackQuotation"] = "BackQuotation";
    /**
     * Language element type
     */
    TokenType["Comment"] = "Comment";
    /**
     * Statement
     */
    TokenType["StatementTerminator"] = "StatementTerminator";
    /**
     * Others
     */
    TokenType["Error"] = "Error";
    /**
     * Left small Bracket
     */
    TokenType["LeftSmallBracket"] = "LeftSmallBracket";
    /**
     * Left small Bracket
     */
    TokenType["RightSmallBracket"] = "RightSmallBracket";
    TokenType["Comma"] = "Comma";
    TokenType["FunctionArguments"] = "FunctionArguments";
})(TokenType || (TokenType = {}));
/**
 *  Token recognition rules
 */
export const TokenReg = {
    [TokenType.StatementTerminator]: /[;]/,
    [TokenType.SingleQuotation]: /['|\']/,
    [TokenType.DoubleQuotation]: /["]/,
    [TokenType.BackQuotation]: /[`]/,
    [TokenType.LeftSmallBracket]: /[(]/,
    [TokenType.RightSmallBracket]: /[)]/,
    [TokenType.Comma]: /[,]/,
};
